content/blog/2018-06-07-you-probably-dont-need-derived-state.md

```diff
@@ -0,0 +1,406 @@
+
+We could use `shouldComponentUpdate` to ensure that our component did not re-render unless props changed. This would fix the simple component showed above. However in practice, components usually accept multiple props, and our component would re-render if any one of them changed— not just email.
+
+Another thing to to keep in mind is that function and object props are often created inline and so will always bypass `shouldComponentUpdate`. For example, what if our component accepted a function to validate the current email address?
+```js
+<EmailInput
+  email={this.props.user.email}
+  validate={this.validateEmail.bind(this)}
+/>
+```
+
+This example binds the validation callback inline and so it will pass a new function prop every time it renders— effectively bypassing `shouldComponentUpdate` entirely. ([Here is a demo that shows that happening.](https://codesandbox.io/s/jl0w6r9w59))
+
+Hopefully it's clear by now why **it is a bad idea to unconditionally mirror props in state**. But what if we were to only update the state when the email prop changes? We'll take a look at that pattern next.
+
+### Anti-pattern: Erasing state when props change
+
+Continuing the example above, we could avoid accidentally erasing state by only updating it when `props.email` changes:
+
+```js
+class EmailInput extends Component {
+  state = {
+    email: this.props.email
+  };
+
+  componentWillReceiveProps(nextProps) {
+    // Any time props.email changes, update state.
+    if (nextProps.email !== this.props.email) {
+      this.setState({
+        email: nextProps.email
+      });
+    }
+  }
+  
+  // ...
+}
+```
+
+> Note
+>
+> Even though the example above shows `componentWillReceiveProps`, the same anti-pattern applies to `getDerivedStateFromProps`.
+
+We've just made a big improvement. Now our component will no longer erase what we've typed when it is re-rendered. At this point, what we have is fairly similar to the [scroll direction example](blog/2018/03/27/update-on-async-rendering.html#updating-state-based-on-props) we mentioned before.
+
+There is still one subtle problem though, and it's probably easiest to illustrate with an example. Let's say our `EmailInput` component is used inside of an "edit profile" form. The first time the form is rendered, the component will display our email address. Let's say we edited the form (including our email address) but then changed our mind and clicked a "reset" button. At this point, we would expect all form fields to return to their initial values— but the `EmailInput` component **will not be reset**. Do you know why?
+
+The reason for this is that `props.email` never actually changed in the above scenario. Both times the "edit profile" form rendered, it would pass the saved email address via props.
```

Comment:
I was with you until this para. Its not clear at all to me that `props.email` "never actually changed. You're just telling me that. I could blame it on poor implementation of the form's reset method. Can you make this clearer?

Reply:
Yeah I think we’ll need codepens for all failing scenarios.

Reply:
I could probably make it clearer by adding the "committed" descriptor in there. Thanks.

Reply:
> The reason for this is that "committed" `props.email` never actually changed in the above scenario. (We never clicked "save".) So when the edit form re-rendered the input, it passed the same email address via props.

Reply:
Here's one I've created for this scenario: https://codesandbox.io/s/mz5w04xpz9

Reply:
And here's one for the Account Manager scenario: https://codesandbox.io/s/mz2lnkjkrx